### PR TITLE
Added file: AbstractSimplicialComplexes.m2

### PR DESCRIPTION
**Pull request by n-m-g**
This package gives a way to work with simplicial complexes implemented as graded lists.  It complements the existing M2 package SimplicialComplexes.m2.

### CODE REVIEW

**Comment by d-torrance** [comment]
I went ahead and fixed it for you, but for future reference, pull requests should target the `development` branch.

**Comment by n-m-g** [comment]
Dear Doug,

OK.  Many thanks.  I’ll try to remember this for future pull-requests.  Do let me know if there are additional issues with any of this.

Best,
Nathan

**Comment by d-torrance** [comment]
You're welcome!

A couple other issues:

* `AbstractSimplicialComplexes` should be added to the file `M2/Macaulay2/packages/=distributed-packages`
* There are some conflicts in `SpectralSequences` between your branch and the `development` branch

**Comment by d-torrance** [comment]
I'm not sure if I'm familiar enough with the code to be of much help...  Here's the big conflict:

```
<<<<<<< HEAD
    C = chainComplex complex L#0; -- By default the ambient simplicial complex is the first element of the list
    maps = apply(#L-1, p -> map(C, chainComplex complex L#(p+1), 
        i -> sub(contract(transpose matrix{faces(i,L#0)}, matrix{faces(i,L#(p+1))}), kk))))
    else (C = truncate(chainComplex complex L#0,1); -- By default the ambient simplicial complex is the first element of the list
    maps = apply(#L-1, p -> map(C, truncate(chainComplex complex L#(p+1),1), 
        i -> sub(contract(transpose matrix{faces(i,L#0)}, matrix{faces(i,L#(p+1))}), kk))))   
=======
    C = complex chainComplex L#0; -- By default the ambient simplicial complex is the first element of the list
    maps = apply(#L-1, p -> map(C, complex chainComplex L#(p+1), 
	    i -> sub(contract(transpose matrix{faces(i,L#0)}, matrix{faces(i,L#(p+1))}), kk))))
    else (C = naiveTruncation(complex chainComplex L#0,1,infinity); -- By default the ambient simplicial complex is the first element of the list
--- the ``patch method" -- truncate a chain complex at a given homological degree 
---- truncate(ChainComplex,ZZ) is now replaced by naiveTruncation(Complex,ZZ,ZZ)
   maps = apply(#L-1, p -> map(C, naiveTruncation(complex chainComplex L#(p+1),1,infinity), 
        i -> sub(contract(transpose matrix{faces(i,L#0)}, matrix{faces(i,L#(p+1))}), kk))));   
>>>>>>> 9300cc5cb (Update SpectralSequences.m2)
```

This looks like it's running into the complexes revamp that @mikestillman has been working on.  The top part is the part of the code that was changed in #3479 , and the bottom part is your new updates.

**Comment by d-torrance** [comment]
There were also a bunch of examples changed in #3479, which is causing a conflict with the documentation, as well.

File: M2/Macaulay2/packages/AbstractSimplicialComplexes.m2 [hunk 1]
```diff
@@ -0,0 +1,832 @@
...
+    "AbstractSimplicialComplexes",
+    Version => "0.1",
+    Date => "24 September 2024",
+    Headline => "AbstractSimplicialComplexes",
+    Authors => {{ Name => "Nathan Grieve", Email => "nathan.m.grieve@gmail.com", HomePage => "https://sites.google.com/view/nathan-grieve"}},
+    AuxiliaryFiles => false,
+    DebuggingMode => false,
+    PackageImports => {"Complexes"},
+    PackageExports => {"Complexes"}
+    )
```
Comment (d-torrance) [inline M2/Macaulay2/packages/AbstractSimplicialComplexes.m2:30]: `newPackage` should have a `Keywords` option that gives a list of one or more strings.  The existing keywords are the headings in the [packages provided with Macaulay2](https://www.macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_packages_spprovided_spwith_sp__Macaulay2.html) documentation page

File: M2/Macaulay2/packages/AbstractSimplicialComplexes.m2 [hunk 2]
```diff
@@ -0,0 +1,832 @@
...
+-- You should have received a copy of the GNU General Public License along with
+-- this program.  If not, see <http://www.gnu.org/licenses/>.
+--------------------------------------------------------------------------------
+
+
+newPackage(
+    "AbstractSimplicialComplexes",
+    Version => "0.1",
+    Date => "24 September 2024",
+    Headline => "AbstractSimplicialComplexes",
```
Comment (d-torrance) [inline M2/Macaulay2/packages/AbstractSimplicialComplexes.m2:24]: The `Headline` should be probably be something more like "abstract simplicial complexes"

**Comment by n-m-g** [comment]
Hi,

This is a comment about the SpectralSequences update?  Yes, this is exactly what we are doing making everything forward compatible at the request of Mike and Greg.  Everything should install and run correctly as far as we are aware.

Best,
Nathan

**Comment by n-m-g** [comment]
Hi,

This seems to be a comment about the spectral sequences update?  If so, yes, we are not familiar with the current status of some of the other complexes updates.  These parts of the code will need to be updated again at a future time.

Best,
Nathan

**Comment by n-m-g** [comment]
Perhaps it's worthwhile mentioning, that both the new package/file AbstractSimplicialComplexes.m2 and the SpectralSequences.m2 "forward compatible" update install and run correctly on Macaulay2, version 1.24.05.

**Comment by mahrud** [comment]
> Perhaps it's worthwhile mentioning, that both the new package/file AbstractSimplicialComplexes.m2 and the SpectralSequences.m2 "forward compatible" update install and run correctly on Macaulay2, version 1.24.05.

There are references to `SpectralSequences2.m2` in the new version, but this file doesn't exist, so I suspect your tests work correct locally because you have some other local files that are not included here.

Also, I noticed all the tests start like:
```m2
TEST ///
restart;
needsPackage "SpectralSequences";
...
```
The `restart` there causes the test to terminate on the first line, so none of the tests have actually been running. This probably should have flagged some sort of error. (Side note: you also don't need to load the package in its own tests, so the `needsPackage` line is also not necessary, but not harmful either)

**Comment by n-m-g** [comment]
Yes.  These tests all run OK but test, the restart should be removed!

**Comment by n-m-g** [comment]
Did you want me to update the SpectralSequences.m2 to remove the "restart" from the test and issue a new pull request, or is this something that you can do on your end?  Do let me know.

**Comment by mahrud** [comment]
Yes, you should remove the restarts, resolve the conflicts with the development branch, fix the issue with missing "SpectralSequence2.m2" files. Doug also requested a few changes above for AbstractSimplicialComplexes that should be implemented.

You don't need to make a new pull request, just implement the changes, commit the changes, and push them to your fork. They will be added here.

**Comment by n-m-g** [comment]
Hi,

Yes, I can do this.  How do I resolve the conflicts with the development branch?  There is no SpectralSequence2.m2 files as we merged that into the existing SpectralSequence.m2 file.  Finally, how do I locate the development branch to fork to; or can I simply use the repo I issued the pull-request from?  Thanks.

Best,
Nathan

**Comment by mahrud** [comment]
The new pull request that you made looks good. Here, I think you should remove SpectralSequences and leave this PR to be just about AbstractSimplicialComplexes.

**Comment by n-m-g** [comment]
Hi,

OK.  I’ll make those changes in the very near future as well.  

Best,
Nathan

**Comment by n-m-g** [comment]
Hi,

AbstrastSimplicialComplexes is good to go now (unless there are other comments); I updated SpectralSequences again so that I have the same file on my fork (but this is now on the development branch now so its fine there as well).

**Comment by d-torrance** [comment]
This branch still has a conflict with `SpectralSequences.m2`.  Since #3552 now exists with the updates to that package, I think it would make sense to remove the SpectralSequences changes from this branch so that it only deals with the new AbstractSimplicialComplexes package.

There are a number of ways to do that, but one way would be:

```
git checkout master
git checkout development -- M2/Macaulay2/packages/SpectralSequences.m2
git add M2/Macaulay2/packages/SpectralSequences.m2
git commit -m "Revert SpectralSequences changes"
git push origin master
```

**Comment by d-torrance** [comment]
Also, a line containing the package name should be added to `M2/Macaulay2/packages/=distributed-packages`.

**Comment by n-m-g** [comment]
Hi,

Many thanks for your follow-up with this.  

Unfortunately, I don’t have git set-up with command line and so I’m just using the web app.  Are you able to fix this on your end, or can I do this online?  If you’d like, we can cancel the existing pull-request and I can make a new one.  

Just let me know what’s easiest.

Best,
Nathan

**Comment by d-torrance** [comment]
Ah, ok.  You should be able to edit `M2/Macaulay2/packages/=distributed-packages` using the GitHub web interface just like you edited the .m2 files, so that's not a problem.

I guess the easiest way to restore SpectralSequences.m2 using the web interface would just be to copy/paste it from its current state on the development branch:

https://raw.githubusercontent.com/Macaulay2/M2/refs/heads/development/M2/Macaulay2/packages/SpectralSequences.m2

**Comment by n-m-g** [comment]
OK.  I’ll try this.  Let’s see if it resolves this issue. 

Again, the current version of the SpectralSequence package is what is there now.   But so long as it is in the correct place now that is fine.  Let me try this “copy paste” solution now.

Best,
Nathan

**Comment by n-m-g** [comment]
Now, it says that my branch is 7 commits ahead which will make new pull requests difficult but I don’t expect any new requests in the near future (aside perhaps from updating existing ones).  Many thanks for your help with this.  Let me know if there is anything else.

Best,
Nathan

**Comment by d-torrance** [comment]
SpectralSequences.m2 is still showing some changes (although there are no more conflicts).  I'll go ahead and remove those commits for you using the command line -- the git history will be cleaner that way without a bunch of SpectralSequences commits that are eventually reverted.

M2/Macaulay2/packages/=distributed-packages still needs to be updated to add the line `AbstractSimplicialComplexes`.

**Comment by d-torrance** [comment]
With the web interface, you should be able to navigate to https://github.com/n-m-g/M2/blob/master/M2/Macaulay2/packages/%3Ddistributed-packages and click the little pencil icon to edit it.  After commiting the changes, it will be automatically added to this PR since it's based on your `master` branch.

**Comment by n-m-g** [comment]
This is great.  Many thanks.  I’ll make changes there in case that anything is needed.  But I think that everything looks good now.  Many thanks again for your help with this.

Best,
Nathan

**Comment by d-torrance** [comment]
We still need the line `AbstractSimplicialComplexes` added to the file `M2/Macaulay2/packages/=distributed-packages`.

**Comment by n-m-g** [comment]
OK.  Thanks.  I understand what you mean now,  I’ve added this now.

Best,
Nathan

**Comment by d-torrance** [comment]
Thanks!

There are still a few conflicts between this branch's copy of `=distributed-packages` and the copy in the `development` branch.  You can fix them using the web interface by clicking the "Resolve conflicts" button which should appear near the bottom of this PR page.  That will bring you to an editor that should show the file with the following at the bottom:

```m2
<<<<<<< master
AbstractSimplicialComplexes
=======
MultigradedImplicitization
Msolve
Permutations
SCMAlgebras
>>>>>>> development
```

Delete the lines with the <<<<'s, ===='s, and >>>>'s, and then click the "Mark as resolved" button.  After that, we should be all set unless any further issues pop up with the package's examples or tests.

**Comment by n-m-g** [comment]
OK.  I've done this now.

**Comment by d-torrance** [comment]
Thanks -- now the builds are running!

There's one error so far:

```m2
../../../../Macaulay2/packages/AbstractSimplicialComplexes.m2:106:15:(2):[62]: error: assignment to protected global variable 'facets', originally defined at ../../../../Macaulay2/packages/Polyhedra/core/globalMethods.m2:121:0:(2):
../../../../Macaulay2/m2/Core.m2:106:28:(1):[61]: --back trace--
../../../../Macaulay2/m2/Core.m2:102:13:(1):[59]: --back trace--
../../../../Macaulay2/m2/Core.m2:113:29:(1):[58]: --back trace--
../../../../Macaulay2/m2/packages.m2:179:8:(1):[57]: --back trace--
```

One possible fix would be to use `facets` from the Polyhedra package.  Something like:

```diff
--- a/M2/Macaulay2/packages/AbstractSimplicialComplexes.m2
+++ b/M2/Macaulay2/packages/AbstractSimplicialComplexes.m2
@@ -31,10 +31,12 @@ newPackage(
     )
 
 export {"AbstractSimplicialComplex", "abstractSimplicialComplex","simplicialChainComplex", "reducedSimplicialChainComplex", "ambientAbstractSimplicialComplexSize",
-    "ambientAbstractSimplicialComplex", "facets", "randomAbstractSimplicialComplex", "randomSubSimplicialComplex",
+    "ambientAbstractSimplicialComplex", "randomAbstractSimplicialComplex", "randomSubSimplicialComplex",
      "inducedSimplicialChainComplexMap","inducedReducedSimplicialChainComplexMap","areEqual", "dimAbstractSimplicialComplex",
     }
 
+exportFrom("Polyhedra", "facets")
+
 -* Code section *-
 
 ---------------------------------------
@@ -103,8 +105,6 @@ select(L,i-> isMaximal(i,L))
 
 --- return the facets of a simplicial complex
 
-facets = method()
-
 facets(AbstractSimplicialComplex) := List => K ->(
     L := flatten(apply(spots K, i-> K_i));
     return listFacets(L)
```

**Comment by d-torrance** [comment]
On the development branch, the Polyhedra package is now one of the preloaded packages, which is causing the conflict.  You can probably reproduce the error on your system by running:

```m2
needsPackage "Polyhedra"
needsPackage "AbstractSimplicialComplexes"
```

**Comment by n-m-g** [comment]
I see.  This makes sense.  So this method needs to overload the existing method, or I can simply rename it in the meantime?  Renaming probably would be most sensible (to avoid other conflicts).

Best,
Nathan

**Comment by n-m-g** [comment]
I’ve renamed “facets” to “abstractSimplicialComplexesFacets” for now; not ideal but this should fix the problem for now.  I’ve committed those changes to my repo.  I trust that this will also update automatically elsewhere?

Best,
Nathan

**Comment by d-torrance** [comment]
Yes, that should work

**Comment by n-m-g** [comment]
Sounds great.  Let’s see if the build fails!

Best,
Nathan

**Comment by d-torrance** [comment]
The builds all worked!  Thanks!

**Comment by n-m-g** [comment]
This is great!

Best,
Nathan

**Comment by n-m-g** [comment]
Hi,

Again, the reason is to keep the package self contained until everything becomes more stable.  This avoids conflicts with items that may be “mutable” in projects that are under development.  My experience with this says that this is the right way to go.

Best,
Nathan

**Comment by mahrud** [comment]
`dim` and `==` are quite stable and have been for as long as I know. `facets` is defined in `Polyhedra` and hasn't changed in years either. The only change is that `Polyhedra` is now preloaded, so the only change needed in your package was to delete `facets = method()` because `Polyhedra` already does that.

**Comment by n-m-g** [comment]
Hi,

Yes, the idea is to have the package independent of Polyhedra.  Perhaps == can be overloaded in due time but for now it’s fine to keep it how it is.  Perhaps also the package could eventually become integrated with Polyhedra but that is something that can be looked into a bit further in due time perhaps.

Best,
Nathan

**Comment by mahrud** [comment]
Sure, if you prefer things the way they are that's fine, though I think there's a misunderstanding because using method name `facets` doesn't mean there's a dependence on Polyhedra, it's just a name.
```m2
i4 : about facets

o4 = {0 => FourierMotzkin :: Finding the facets of the cyclic polytope}
     {1 => FourierMotzkin :: Finding the facets of the permutahedron  }
     {2 => Polyhedra :: facets                                        }
     {3 => Polyhedra :: facets(Cone)                                  }
     {4 => Polyhedra :: facets(Polyhedron)                            }
     {5 => SimplicialComplexes :: facets(SimplicialComplex)           }
     {6 => SRdeformations :: facets                                   }
     {7 => SRdeformations :: facets(Complex)                          }
     {8 => SRdeformations :: facets(List)                             }
```
Similarly, tons of packages use `dim`:
```m2
i3 : about dim

o3 = {0 => CellularResolutions :: dim(Cell)                        }
     {1 => CellularResolutions :: dim(CellComplex)                 }
     {2 => Chordal :: dim(ChordalNet)                              }
     {3 => Chordal :: dim(ChordalNetChain)                         }
     {4 => CodingTheory :: dim(LinearCode)                         }
     {5 => CoincidentRootLoci :: dim(CoincidentRootLocus)          }
     {6 => GradedLieAlgebras :: dim(List,ExtAlgebra)               }
     {7 => GradedLieAlgebras :: dim(List,LieAlgebra)               }
     {8 => GradedLieAlgebras :: dim(List,VectorSpace)              }
     {9 => GradedLieAlgebras :: dim(ZZ,ExtAlgebra)                 }
     {10 => GradedLieAlgebras :: dim(ZZ,LieAlgebra)                }
     {11 => GradedLieAlgebras :: dim(ZZ,VectorSpace)               }
     {12 => GradedLieAlgebras :: dim(ZZ,ZZ,ExtAlgebra)             }
     {13 => GradedLieAlgebras :: dim(ZZ,ZZ,LieAlgebra)             }
     {14 => GradedLieAlgebras :: dim(ZZ,ZZ,VectorSpace)            }
     {15 => InvariantRing :: dim(GroupAction)                      }
     {16 => KustinMiller :: dim(Face)                              }
     {17 => LieTypes :: dim(LieAlgebra)                            }
     {18 => LieTypes :: dim(LieAlgebraModule)                      }
     {19 => LocalRings :: dim(LocalRing)                           }
     {20 => Macaulay2Doc :: dim                                    }
     {21 => Macaulay2Doc :: dim(FractionField)                     }
     {22 => Macaulay2Doc :: dim(GaloisField)                       }
     {23 => Macaulay2Doc :: dim(Ideal)                             }
     {24 => Macaulay2Doc :: dim(InexactField)                      }
     {25 => Macaulay2Doc :: dim(Module)                            }
     {26 => Macaulay2Doc :: dim(MonomialIdeal)                     }
     {27 => Macaulay2Doc :: dim(PolynomialRing)                    }
     {28 => Macaulay2Doc :: dim(ProjectiveHilbertPolynomial)       }
     {29 => Macaulay2Doc :: dim(QuotientRing)                      }
     {30 => Macaulay2Doc :: dim(Ring)                              }
     {31 => MultiprojectiveVarieties :: dim(MultiprojectiveVariety)}
     {32 => NAGtypes :: dim(Ambient)                               }
     {33 => NAGtypes :: dim(DualSpace)                             }
     {34 => NAGtypes :: dim(NumericalVariety)                      }
     {35 => NAGtypes :: dim(PolySpace)                             }
     {36 => NAGtypes :: dim(SlicingVariety)                        }
     {37 => NAGtypes :: dim(WitnessSet)                            }
     {38 => NAGtypes :: dim(WSet)                                  }
     {39 => NormalToricVarieties :: dim(NormalToricVariety)        }
     {40 => OldPolyhedra :: dim(Cone)                              }
     {41 => OldPolyhedra :: dim(Fan)                               }
     {42 => OldPolyhedra :: dim(PolyhedralComplex)                 }
     {43 => OldPolyhedra :: dim(Polyhedron)                        }
     {44 => Polyhedra :: dim(PolyhedralObject)                     }
     {45 => Schubert2 :: dim(AbstractVariety)                      }
     {46 => Schubert2 :: dim(AbstractVarietyMap)                   }
     {47 => SchurRings :: dim(List,SchurRingElement)               }
     {48 => SchurRings :: dim(SchurRingElement)                    }
     {49 => SchurRings :: dim(Thing,SchurRingElement)              }
     {50 => SimplicialComplexes :: dim(SimplicialComplex)          }
     {51 => SparseResultants :: dim(MultidimensionalMatrix)        }
     {52 => SRdeformations :: dim(Complex)                         }
     {53 => SRdeformations :: dim(Face)                            }
     {54 => SRdeformations :: dim(Face,Complex)                    }
     {55 => SRdeformations :: dim(Face,PolynomialRing)             }
     {56 => SRdeformations :: dim(FirstOrderDeformation)           }
     {57 => TriangularSets :: dim(TriaSystem)                      }
     {58 => Tropical :: dim(TropicalCycle)                         }
     {59 => Varieties :: dim(AffineVariety)                        }
     {60 => Varieties :: dim(ProjectiveVariety)                    }
```

**Comment by mahrud** [comment]
For future reference, I also want to point out the [Package Writing Style Guide](https://github.com/Macaulay2/M2/wiki/Package-Writing-Style-Guide) which states:
> Names representing methods [...] should not include the name of the type of object expected as argument, since the idea of such methods is that they are mathematical abstractions that act on a variety of types of mathematical object.

**Comment by n-m-g** [comment]
Hi,

There is no misunderstanding here.  The point is that overloading methods in a package sometime requires adding additional options etc. (which is why the previous build didn’t work).  It’s easy to overload things if needed but also OK not to.   Again, it might be interesting to explore some integration with Polyhedra eventually but this is not something that has been done yet.

Best,
Nathan

**Comment by n-m-g** [comment]
Thanks for this reminder!  I will keep in mind for future reference; I don’t think that my naming conventions conflict/infringe too much with those conventions.  Again, one has to be careful when overloading (so as too avoid undue confusion) as I’m sure you are aware.

Best,
Nathan

**Comment by mahrud** [comment]
`dimAbstractSimplicialComplex` includes the name of the type of object expected as input, which is `AbstractSimplicialComplex`, so this is explicitly against the style guide. Neither `dim`, `facets`, nor `==` accept any options either, so that can't be the reason why the previous build didn't work.

Again, I'm not suggesting any integration with Polyhedra whatsoever, and you're welcome to have your package as is, but I am not aware of issues with overloading these methods or any confusion that can be caused as a result.

**Comment by mahrud** [comment]
We moved `cone` and `rays` to `shared.m2` for the same reason, as well as `isEmpty` and `isVeryAmple` recently.

I believe SimplicialComplexes exports Polyhedra because they both use terms like `skeleton`, `facets`, `vertices`, etc., but it doesn't actually depend on Polyhedra or integrate with it in any way (at least as far as I can tell).

**Comment by n-m-g** [comment]
Many thanks for these additional comments.  Again, these are all reasons to keep this package mostly self contained (aside from using the Complexes package) at present.  Perhaps these are all items that we can discuss offline (unless it’s still of interest to continuing discussing them here).

Best,
Nathan

**Comment by mahrud** [comment]
Yes, some changes definitely need to be made before the next release, for instance the keywords added should be chosen from existing keywords which are the headlines on [this page](https://www.macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_packages_spprovided_spwith_sp__Macaulay2.html):
```m2
    Keywords => {"Simplicial complexes", "Simplicial chain complexes", "Random simplicial complexes"}
```

We can discuss the other issues on Zulip if you prefer, but the reasons you mentioned so far (e.g. optional argument conflicts or potential dependency on unstable package) don't apply to `dimAbstractSimplicialComplex`.

**Comment by n-m-g** [comment]
Sure.  I am happy to make changes as required to have everything approved.  Yes, "dim" is another one of those overloaded methods that gets used in different senses at times.  I'm not sure what you mean by choosing keywords from existing key words.  In preparing this package I just followed what we did in the spectral sequences package.  We didn't choose any keywords etc.  Just let me know what changes you would like me to make and I can try to do this prior to the next release.

**Comment by n-m-g** [comment]
Do you mean what "subject area keywords" it should fall into?  For that, I would use "homological algebra" as primary and maybe "combinatorics" and/or "combinatorial commutative algebra" as secondary "subject area keywords".

**Comment by d-torrance** [comment]
The `Keywords` option to `newPackage` should ideally be chosen from the keywords for existing packages, which can be seen as the headings at the [packages provided with Macaulay2](https://www.macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_packages_spprovided_spwith_sp__Macaulay2.html) documentation page.

I went ahead at took the liberty of changing the keyword to "Combinatorial Commutative Algebra" (to match the non-abstract SimplicialComplexes package) in #3535.

**Comment by n-m-g** [comment]
OK.  This is great.  Many thanks!  Do let me know if there is anything that I need to do prior to the next release.

Best,
Nathan

**Comment by n-m-g** [comment]
If it's not too late, I just did the small edits to overload == and dim as requested as well.  I've just committed those changes.

**Comment by d-torrance** [comment]
Sure! Feel free to open a new pull request

**Comment by n-m-g** [comment]
This is great.  I’ve just done that now.

Best,
Nathan

**Pull request merged**